File: Week01/wordSearch.js
// @ts-nocheck
//Given an m x n grid of characters board and a string word, return true if word exists in the grid.

//The word can be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once.

//Example 1:
// Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCCED"
//Output: true

//Example 2:
//Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "SEE"
//Output: true

//Example 3:
//Input: board = [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], word = "ABCB"
//Output: false

// Constraints:

// m == board.length (number of lines)
// n = board[i].length (number of characters for each line)
// 1 <= m, n <= 6 (on line minimal and no more than 6 characters per line)
// 1 <= word.length <= 15 (word can have less than 16 characters )
// board and word consists of only lowercase and uppercase English letters.

//Follow up: Could you use search pruning to make your solution faster with a larger board?

//word: "ABCCED"
//["A*", "B*", "C*", "E"]
//["S",  "F",  "C*", "S"]
//["A",  "D*", "E*", "E"]

//if line is equal to 0 do not test UP
//if line is equal to the length - 1 do not test down
//if colum is equal to the length - 1 do not test down 

var exist = function (board, word) {
   let totLin = board.length;
   let totCol = board[0].length;
   let totChar = word.length;
   let currentChar = 0;
   let lin = 0;
   let col = 0;

   if (totLin > totLin * totCol)
      return false;

   while (lin < totLin && col < totCol) {
      if (word[currentChar] === board[lin][col]) {
         if (findPath(0, lin, col))
            return true;
         else
            moveForward();
      } else {
         moveForward();
      }
   }
   return false;

   function findPath(position, foundLin, foundCol) {
      const left = -1;
      const right = +1;
      const up = -1;
      const down = +1;

      if (totChar === position)
         return true;

      if (foundLin < 0 || foundCol < 0 || foundLin >= totLin || foundCol >= totCol)
         return false;

      if (board[foundLin][foundCol] === word[position]) {
         let tempChar = board[foundLin][foundCol];
         board[foundLin][foundCol] = '#';

         let result = findPath(position + 1, foundLin + down, foundCol) |
            findPath(position + 1, foundLin + up, foundCol) |
            findPath(position + 1, foundLin, foundCol + left) |
            findPath(position + 1, foundLin, foundCol + right);

         board[foundLin][foundCol] = tempChar;
         return result;

      } else
         return false;
   }

   function moveForward() {
      if (col < totCol - 1) {
         col++;
      } else if (lin < totLin) {
         lin++;
         col = 0;
      }
   }
};



//console.log(exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "ABCCED"));
//console.log(exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "SEE")); //true
//console.log(exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "ABCB")); //false
//console.log(exist([["a", "b"], ["c", "d"]], "abdc")); //true
//console.log(exist([["a", "b"], ["c", "d"]], "abcd")); //False
//console.log(exist([["a", "b"]], "ab")); //true
//console.log(exist([["a", "b"], ["c", "d"]], "acdb")); //true
//console.log(exist([["a", "a"]], "a")); //true
//console.log(exist([["C", "A", "A"], ["A", "A", "A"], ["B", "C", "D"]], "AAB")); //true
console.log(exist([["A", "B", "C", "E"], ["S", "F", "E", "S"], ["A", "D", "E", "E"]], "ABCESEEEFS")); //true
//console.log(exist([["a", "a", "a", "a"], ["a", "a", "a", "a"], ["a", "a", "a", "a"]], "aaaaaaaaaaab")); //true

File: Patterns/BreathFirstSearch/Python/word_search.py
class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        tot_lin = len(board)
        tot_col = len(board[0])
        tot_char = len(word)
        current_char, lin, col = 0, 0, 0

        if tot_lin > tot_lin * tot_col:
            return False

        while lin < tot_lin and col < tot_col:
            if word[current_char] == board[lin][col]:
                if self.find_path(board, word, tot_char, tot_lin, tot_col, 0, lin, col):
                    return True
                else:
                    self.move_forward(col, tot_col, lin, tot_lin)
            else:
                self.move_forward(col, tot_col, lin, tot_lin)

        return False

    def find_path(self, board, word, tot_char, tot_lin, tot_col, position, found_lin, found_col) -> bool:
        left, right, up, down = -1, +1, -1, +1

        if tot_char == position:
            return True

        if found_lin < 0 or found_col < 0 or found_lin >= tot_lin or found_col >= tot_col:
            return False

        if board[found_lin][found_col] == word[position]:
            temp_char = board[found_lin][found_col]
            board[found_lin][found_col] = '#'

            result = (self.find_path(board, word, tot_char, tot_lin, tot_col, position + 1, found_lin + down, found_col) |
                      self.find_path(board, word, tot_char, tot_lin, tot_col, position + 1, found_lin + up, found_col) |
                      self.find_path(board, word, tot_char, tot_lin, tot_col, position + 1, found_lin, found_col + left) |
                      self.find_path(board, word, tot_char, tot_lin, tot_col, position + 1, found_lin, found_col + right))

            board[found_lin][found_col] = temp_char
            return result
        else:
            return False

    def move_forward(self, col, tot_col, lin, tot_lin):
        if col < tot_col - 1:
            col += 1
        elif lin < tot_lin:
            lin += 1
            col = 0


# print(Solution().exist([["A", "B", "C", "E"], ["S", "F", "C", "S"],
#       ["A", "D", "E", "E"]], "ABCCED"))
print(Solution().exist([["A", "B", "C", "E"], [
      "S", "F", "C", "S"], ["A", "D", "E", "E"]], "SEE"))
